scikeras/wrappers.py

```diff
@@ -314,14 +335,36 @@ def _validate_data(self, X, y=None, reset=True):
         out : {ndarray, sparse matrix} or tuple of these
             The validated input. A tuple is returned if `y` is not None.
         """
+        # default to TFs backend float type
+        # instead of always float64 (sklearns default)
+        tf_backend_dtype = np.dtype(tf.keras.backend.floatx())
+
+        def get_dtype(arr) -> np.dtype:
+            """Helper function to determine the correct
+            dtype for an input `arr`.
+            """
+            output_dtype = tf_backend_dtype
+            if isinstance(arr, np.ndarray) and arr.dtype.kind != "O":
+                output_dtype = arr.dtype
+            else:
+                arr_dtype = np.array(arr).dtype
+                if arr_dtype.kind != "O":
+                    # list of arrays with equal dtype
+                    output_dtype = arr_dtype
+            return output_dtype
+
         if y is not None:
             X, y = check_X_y(
                 X,
                 y,
                 allow_nd=True,  # allow X to have more than 2 dimensions
                 multi_output=True,  # allow y to be 2D
+                dtype=None,
+            )
+            y = check_array(
+                y, ensure_2d=False, allow_nd=False, dtype=get_dtype(y)
             )
-        X = check_array(X, allow_nd=True, dtype=["float64", "int"])
+        X = check_array(X, allow_nd=True, dtype=get_dtype(X))
```

Comment:
What's the function `get_dtype` for? Why not pass `dtype="numeric"` to `check_array`? At first glance it looks like `check_array` does the same thing as `get_dtype`:

> **dtype: <i>string, type, list of types or None (default=”numeric”)</i>**
> Data type of result. If None, the dtype of the input is preserved. If “numeric”, dtype is preserved unless array.dtype is object. If dtype is a list of types, conversion on the first type is only performed if the dtype of the input is not in the list.
>
> —[sklearn.utils.check_array docs][1]

It looks like the outputs of `check_array(X, dtype=get_dtype(X))` and `check_array(X, dtype="numeric")` only differ when `X` has an object dtype (in which case they output objects of  `backend.floatx()` and `float64` respectively). Is that accurate?

[1]:https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_array.html

Reply:
Exactly. The outcome we want is:
- object input: float32 output
- numeric input: same type as output